klippa/src/lib.rs

```diff
@@ -7,67 +7,216 @@ mod parsing_util;
-        let maxp = font.maxp().expect("Error reading maxp table");
-        let maxp = Maxp::from_table_ref(&maxp);
-        let gids: IntSet<GlyphId> = gids
-            .iter()
-            .filter(|gid| gid.to_u16() < maxp.num_glyphs)
-            .collect();
-        let num_glyphs = gids.len() as u16;
+        let mut this = Plan::empty();
+        // store font's num of glyphs in the plan
+        this.font_num_glyphs = get_font_num_glyphs(font);
+
+        this.populate_unicodes_to_retain(input_gids, input_unicodes, font);
+        this.populate_gids_to_retain(font);
+        this.num_output_glyphs = this.glyphset.len() as u16;
 
         // compute new h_metrics
         let hmtx = font.hmtx().expect("Error reading hmtx table");
         let hmtx = Hmtx::from_table_ref(&hmtx);
-        let new_h_metrics = compute_new_num_h_metrics(&hmtx, &gids);
+        this.num_h_metrics = compute_new_num_h_metrics(&hmtx, &this.glyphset);
 
-        Self {
-            glyph_ids: gids,
-            num_h_metrics: new_h_metrics,
-            num_output_glyphs: num_glyphs,
-        }
+        this
     }
-}
 
-fn populate_unicodes_to_retain(
-    input_gids: &IntSet<GlyphId>,
-    input_unicodes: &IntSet<u32>,
-    font: &FontRef,
-) -> IntSet<GlyphId> {
-    let mut gids = IntSet::empty();
-    gids.insert(GlyphId::new(0_u16));
-    gids.extend(input_gids.iter());
-
-    let cmap = font.cmap().expect("Error reading cmap table");
-    for cp in input_unicodes.iter() {
-        match cmap.map_codepoint(cp) {
-            Some(gid) => {
-                gids.insert(gid);
+    pub fn populate_unicodes_to_retain(
+        &mut self,
+        input_gids: &IntSet<GlyphId>,
+        input_unicodes: &IntSet<u32>,
+        font: &FontRef,
+    ) {
+        let charmap = font.charmap();
+        if input_gids.is_empty() && input_unicodes.len() < self.font_num_glyphs {
+            let cap = input_unicodes.len();
+            self.unicode_to_new_gid_list.reserve(cap);
+            self.codepoint_to_glyph.reserve(cap);
+            //TODO: add support for subset accelerator?
+
+            for cp in input_unicodes.iter() {
+                match charmap.map(cp) {
+                    Some(gid) => {
+                        self.codepoint_to_glyph.insert(cp, gid);
+                        self.unicode_to_new_gid_list.push((cp, gid));
+                    }
+                    None => {
+                        continue;
+                    }
+                }
             }
-            None => {
-                continue;
+        } else {
+            //TODO: add support for subset accelerator?
+            let charmap = font.charmap();
+            let cmap_unicodes = charmap.mappings().map(|t| t.0).collect::<IntSet<u32>>();
+            let unicode_gid_map = charmap.mappings().collect::<HashMap<u32, GlyphId>>();
+
+            let mut vec_cap = input_gids.len() + input_unicodes.len();
+            vec_cap = vec_cap.min(cmap_unicodes.len());
```

Comment:
this is a huge nit from me, so please forgive me, but: I find in general if you can write something in such a way as to avoid needing mutation it's worth doing so, unless it's much more awkward.

The basic rationale is that when reading code, anytime you see `mut` you suddenly need to keep track of the fact that this value might be changed at any point from there forwards, which is an extra burden to understanding.

In this particular case, there's a very simple alternative which is common in rust (but is sometimes discouraged in other languages) and that is to reuse a binding. Specifically, you can write

```suggestion
            let vec_cap = input_gids.len() + input_unicodes.len();
            let vec_cap = vec_cap.min(cmap_unicodes.len());
```

where the second binding is shadowing the first.

There's a good section on this [in the Rust docs](https://doc.rust-lang.org/1.79.0/book/ch03-01-variables-and-mutability.html#shadowing).

Reply:
thanks! This is very useful